services/brig/src/Brig/User/Auth.hs

```diff
@@ -252,12 +252,34 @@ isPendingActivation ident = case ident of
             Just SSOIdentity {} -> False -- sso-created users are activated immediately.
             Nothing -> True
 
+-- | Validate a list of (User/LH) tokens potentially with an associated access token.
+--   If there are multiple valid cookies, we try all of them. When an access token is
+--   given, we perform the usual checks.
+--   If multiple cookies are given and several are valid, we return the first valid one.
 validateTokens ::
+  ZAuth.TokenPair u a =>
+  [ZAuth.Token u] -> -- FUTUREWORK: This should be a NonEmpty
+  Maybe (ZAuth.Token a) ->
+  ExceptT ZAuth.Failure AppIO (UserId, Cookie (ZAuth.Token u))
+validateTokens [] _ = throwE ZAuth.Invalid
+validateTokens (ut : []) at = validateToken ut at
+validateTokens uts at = do
```

Comment:
Is this function good enough?

Currently it will fail if _any_ of the cookies is malformed.  But succeed if they are all well-formed but only a few of them is  valid.

Don't we also want to support the usecase of   ` [ malformed, valid, malformed]` too?


Say we make a breaking change in the `zuid` format; then this function would handle that gracefully in the future; which sounds very useful to me.

Reply:
We discussed it f2f too: I actually said back theb that "if the cookie is not parseable as Zauth, then it's dropped altogether when parsing the header in wai-routing".

So I added a test and... realised I was wrong! I fixed it now with https://github.com/wireapp/wire-server/pull/1102/commits/6841dff9e5615c0816f9681743f4b6c45072197b and realised why during our call I wanted to use `catMaybes` and `map` :P `traverse` fails too early